util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,284 @@
+	udpClientCmd = []string{"-c", "{destIP}", "-u", "-f", "K", "-l", "20", "-b", "1M", "-e", "-i", "1", "-t", "{duration}"}
+	udpServerCmd = []string{"-s", "-f", "K", "-u", "-e", "-i", "{duration}", "-P", "{numCls}"}
+	tcpServerCmd = []string{"-s", "-f", "K", "-i", "{duration}", "-P", "{numCls}"}
+	tcpClientCmd = []string{"-c", "{destIP}", "-f", "K", "-l", "20", "-b", "1M", "-i", "1", "-t", "{duration}"}
+	// Siege command args:
+	// -d1 : random delay between 0 to 1 sec
+	// -t<duration>S : run test for <duration> seconds
+	// -c1 : one concurrent user
+	httpClientCmd = []string{"http://" + "{destIP}" + ":" + httpPort + "/test", "-d1", "-t" + "{duration}" + "S", "-c1"}
+)
+
+// Start worker
+func (w *Worker) Start() {
+	w.resultStatus = make(chan string)
+	w.listenToController()
+}
+
+func (w *Worker) listenToController() {
+	h := map[string]func(http.ResponseWriter, *http.Request){"/startTCPServer": w.StartTCPServer,
+		"/startTCPClient": w.StartTCPClient, "/startUDPServer": w.StartUDPServer, "/startUDPClient": w.StartUDPClient,
+		"/startHTTPServer": w.StartHTTPServer, "/startHTTPClient": w.StartHTTPClient, "/metrics": w.Metrics}
+	w.startListening(workerListenPort, h)
+	klog.Info("Started listening to Server")
+}
+
+func (w *Worker) startListening(port string, handlers map[string]func(http.ResponseWriter, *http.Request)) {
+	klog.Info("About to listen to controller requests")
+	for p, h := range handlers {
+		http.HandleFunc(p, h)
+	}
+	err := http.ListenAndServe(":"+port, nil)
+	if err != nil {
+		klog.Fatalf("Failed starting http server for port: %v, Error: %v", port, err)
+	}
+}
+
+// Metrics returns the metrics collected
+func (w *Worker) Metrics(res http.ResponseWriter, req *http.Request) {
+	var reply metricResponse
+	klog.Info("In metrics")
+	var stat string
+	select {
+	case stat = <-w.resultStatus:
+		if stat != "OK" {
+			klog.Error("Error collecting metrics:", stat)
+			reply.Error = "metrics collection failed:" + stat
+			w.createResponse(reply, &res)
+			return
+		}
+		klog.Info("Metrics collected")
+	default:
+		klog.Info("Metric collection in progress")
+		reply.Error = "metric collection in progress"
+		w.createResponse(reply, &res)
+		return
+	}
+	o, err := util.ParseResult(w.tstProtocol, w.result, w.tstDuration)
+	if err != nil {
+		reply.Error = err.Error()
+		w.createResponse(reply, &res)
+		return
+	}
+	reply.Result = o
+	//startime and other variables below can be removed in future, for debugging during initial runs
+	stStr := strconv.FormatInt(w.startedAt, 10)
+	ftStr := strconv.FormatInt(w.futureTime, 10)
+	dtStr := strconv.FormatInt(w.futureTime-w.startedAt, 10)
+	reply.WorkerStartTime = "StartedAt:" + stStr + " FutureTime:" + ftStr + " diffTime:" + dtStr
+	w.createResponse(reply, &res)
+}
+
+func (w *Worker) createResponse(resp interface{}, wr *http.ResponseWriter) {
+	klog.Info("Inside create resp")
+	(*wr).Header().Set("Content-Type", "application/json")
+	(*wr).WriteHeader(http.StatusOK)
```

Comment:
Actually, sending StatusOK if the response has an error seems wrong to me.

The createResponse shouldn't be type-agnostic and should send the correct http code.

Reply:
Modified.